prov/verbs: update man page with new env variables

a-ilango opened the pull request:
- prov/verbs: Update man page with new env variables.

shefty commented on man/fi_verbs.7.md:
```diff
@@ -170,6 +170,14 @@ The verbs provider checks for the following environment variables.
+: Maximum number of cache entries (default: 4096)
+
+*FI_VERBS_MR_CACHE_LAZY_SIZE*
+: Minimum size of lazy deregistration (default: 0)
```
Please expand on the documentation.  Unless someone is familiar with verbs and what a MR cache is, they won't really know what these fields do or the affect that they have.  I would describe what a MR cache actually is, and provide more details on what a 'lazy size' is, in particular.

gladkovdmitry17 commented on prov/rxm/src/rxm_init.c:
```diff
@@ -46,7 +46,9 @@ int rxm_info_to_core(uint32_t version, const struct fi_info *hints,
+	     !(hints->domain_attr->mr_mode & (FI_MR_BASIC | FI_MR_SCALABLE))))
 		core_info->domain_attr->mr_mode |= FI_MR_LOCAL;
 	else
 		core_info->mode |= FI_LOCAL_MR;
```
Should we remove other Basic MR flags (`FI_MR_VIRT_ADDR`, `FI_MR_ALLOCATED`, and `FI_MR_PROV_KEY`) in case of `mr_mode & FI_MR_BASIC` provided? Since setting of these flags are equivalent of setting of `FI_MR_BASIC` flag

a-ilango replied:
If the app is providing FI_MR_BASIC it should not specify any other MR mode bits. So, removing them may not be needed.

gladkovdmitry17 replied:
Ok. Maybe the right way is to not add `FI_MR_PROV_KEY` during opening core provider domain

a-ilango commented:
> but we still need to evaluate why it doesn't work for the RxM/verbs/MSG provider when an user passes FI_MR_BASIC MR mode

Thanks. I'm looking into the issues.

a-ilango commented:
@shefty I've updated the man page with MR cache details. Also fixed an issue with util MR mode handling. Please take a look.
@gladkovdmitry17 I've updated MR mode handling in rxm provider. Please let me know if the tests run fine.

gladkovdmitry17 commented:
@a-ilango, it works fine for me! Feel free to merge this patch
Thanks.

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -1182,15 +1185,9 @@ void ofi_alter_info(struct fi_info *info, const struct fi_info *hints,
-		if (FI_VERSION_LT(api_version, FI_VERSION(1, 5))) {
-			if (info->domain_attr->mr_mode & FI_MR_LOCAL)
-				info->mode |= FI_LOCAL_MR;
-		}
```
The reason info->mode was altered here is so that a call like alter_domain_attr() doesn't need to alter anything other than domain attributes, which ends up looking like a unexpected side effect.

a-ilango replied:
Agreed. I wanted to keep the checks in one place but I guess having the same checks here wouldn't hurt and it helps to  avoid the side effect.

shefty commented on man/fi_verbs.7.md:
```diff
@@ -88,6 +88,20 @@ for restrictions. It can be turned off by setting the FI_FORK_UNSAFE environment
+The verbs provider features a memory registration cache. This speeds up memory
+registration calls from applications by caching registrations of frequently used
+memory regions. It also has the option of lazy memory registration cache where
+registrations of memory regions only above a certain size get cached. The user
```
Slight wording tweak starting at "It also has the option...":
Cached registrations are lazily unregistered, which allows registration calls of the same region to only go through the registration process once.  Additional controls allow caching only registrations of memory regions under a specified size.

a-ilango replied:
@gladkovdmitry17 From the code it seems when `FI_VERBS_MR_CACHE_LAZY_SIZE` is set, memory registrations for buffers only above a certain size would be cached:
https://github.com/ofiwg/libfabric/blob/2640bcca250548feb713cf0a7add2e5075fa5d03/prov/verbs/src/verbs_mr.c#L332

Is that the intended behavior? @shefty and I discussed about this and it seems we should cache memory registration until a certain size and not the other way around. An application might use small buffers many times than a larger one.

gladkovdmitry17 replied:
Yes, that is the intended behavior. The larger the buffer, the more expensive it is to register, isn't it? From our experiments it should be so

shefty replied:
Registration costs are related to the number of pages being registered, but the cost of registration relative to the cost of the transfer is greater for smaller messages.  That is, I believe you'll see a bigger drop in performance doing: register 4k + send 4k + unregister 4k, than you will if the buffer size were 1 GB.

Pinning large regions will impact how much memory is available across the system, so it affects all processes.  So, for large transfers, I don't think you want to maintain pinning that memory longer than is necessary.  I also wouldn't expect an app to re-use a 1 GB buffer, for example, versus reusing a 4k buffer.

This is why I would expect to see limitations placed on how large a region can be and still be cached, as well as a possible limit on the total amount of memory cached at once.

gladkovdmitry17 replied:
> Registration costs are related to the number of pages being registered,
> but the cost of registration relative to the cost of the transfer is greater for smaller messages.

Yes, I agree.

> Pinning large regions will impact how much memory is available across the system

Sure, but our goal is to reduce cost that we need to pay when sending large regions. Unfortunately, this scheme is not good, but it works.

Take into consideration our discussions., I'd say the following:
The idea of having both `FI_VERBS_MR_CACHE_LAZY_SIZE` and `FI_VERBS_MR_CACHE_SIZE` variables is  improper.
Why we should count how many entries were registered and limit size from which  provider should cache regions?  let's set  the threshold - how much memory can be cached?
For example, 1 GB. 
User can register 1 GB region and other regions won't be cached. But as an opposite example - user can register 1024 regions of 1024 MB... and etc. But there is limit - 1 GB. It will be simply controlled, we will have a variable that will contain size (how much memory is registered) and it will be limited by threshold.

 I guess, my idea is clear :) What do you think? Sorry that this idea didn't come to me before :)

shefty replied:
I like the idea of having a single threshold that will limit the total amount of memory registered by the cache.

gladkovdmitry17 replied:
> I like the idea of having a single threshold that will limit the total amount of memory registered by the cache.

Thanks, Sean
So, I'll do that

shefty commented on man/fi_verbs.7.md:
```diff
@@ -170,6 +184,14 @@ The verbs provider checks for the following environment variables.
+: Maximum number of cache entries (default: 4096)
+
+*FI_VERBS_MR_CACHE_LAZY_SIZE*
+: Minimum memory region size for lazy deregistration (default: 0)
```
Please verify if this is a minimum size for caching or a maximum size.  That is, are regions larger than this size cached or not cached.

j-xiong replied:
The terminology here is somewhat confusing, with "CACHE_SIZE" means the # of entries while "CACHE_LAZY_SIZE" means a threshold of buffer size above which the registration is cached. One of both of them may need to be renamed to present a clearer interface.

gladkovdmitry17 replied:
@j-xiong `FI_VERBS_MR_CACHE_SIZE` -> `FI_VERBS_MR_CACHE_ENT_NUM` or something like that?

j-xiong replied:
@gladkovdmitry17 That's in the right direction I think. We could completely avoid using SIZE here. For example: `FI_VERBS_MR_CACHE_SIZE` -> `FI_VERBS_MR_CACHE_ENTRIES`, `FI_VERBS_MR_CACHE_LAZY_SIZE` -> `FI_VERBS_MR_CACHE_THRESHOLD`.

a-ilango commented:
I've moved the commits related to MR mode to #3768

hppritcha commented:
bot:lanl:retest

a-ilango commented:
@gladkovdmitry17 any update on changing the behavior of variables?

If we can get the updates before 1.6 release, it would be good. Otherwise, we can document the current behavior in the man pages with the disclaimer that it's experimental and might change in the future.

gladkovdmitry17 commented:
@a-ilango The #3874 PR changes the behavior of variables. Could you take a look, please?

a-ilango commented:
Thanks @gladkovdmitry17 for the quick response! I've updated this PR with the new env variables.

shefty commented:
Is this PR up to date now with the latest merge?

a-ilango commented:
> Is this PR up to date now with the latest merge?

Yes, this is up to date.
